File: src/my_py_pkg/my_py_pkg/count_until_server_minimal.py
#!/usr/bin/env python3
import time

import rclpy
from rclpy.action import ActionServer, GoalResponse
from rclpy.action.server import ServerGoalHandle
from rclpy.node import Node

from my_robot_interfaces.action import CountUntil


class CountUntilServerNode(Node):
    def __init__(self):
        super().__init__("count_until_server")
        self.count_until_server_ = ActionServer(
            self, # node to link action server to. topics and serices use `self.create...()`. here the object (self) is used instead.
            CountUntil, 
            "count_until", # action name
            goal_callback=self.goal_callback,
            execute_callback=self.execute_callback)
        
    # Every new received goal will be processed here first    
    def goal_callback(self, goal_request: CountUntil.Goal):
        self.get_logger().info("Received a goal")
        if goal_request.target_number <= 0:
            self.get_logger().info("Rejecting the goal, target number must be positive")
            return GoalResponse.REJECT
        self.get_logger().info("Accepting the goal")
        return GoalResponse.ACCEPT
        
    # If a goal has been accepted, it will then be executed in this callback
    # After we are done with the goal execution we set a final state and return the result
    def execute_callback(self, goal_handle: ServerGoalHandle):
        target_number = goal_handle.request.target_number
        delay = goal_handle.request.delay
        result = CountUntil.Result()
        counter = 0

        self.get_logger().info("Executing the goal")
        for i in range (target_number):
            counter += 1
            self.get_logger().info(str(counter))
            time.sleep(delay)
            
        goal_handle.succeed() # we'l only use this for this simple example
        result.reached_number = counter
        return result
    
def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServerNode()
    rclpy.spin(node)
    rclpy.shutdown()
    
if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/count_until_server_minimal.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/count_until.hpp"

using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ServerGoalHandle<CountUntil>;
using namespace std::placeholders;


// the C++ syntax is stricter than Python for actions
// you have to provide three callbacks (even if you don’t want to use them all)
// Goal callback, Cancel callback, and Execute callback
// Excecute callback is called the handle accepted callback in C++, but we named it execute callback to make the code similar to the Python one. 
// In this callback, we execute goals that have been accepted.
// this is intednded to be a minimal example of action server in C++
// the full mehtod to handle cancel mechanism is implemented later
class CountUntilServerNode : public rclcpp::Node
{
public:
    CountUntilServerNode() : Node("count_until_server")
    {
        count_until_server_ = rclcpp_action::create_server<CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServerNode::goalCallback, this, _1, _2),
            std::bind(&CountUntilServerNode::cancelCallback, this, _1),
            std::bind(&CountUntilServerNode::executeCallback, this, _1)
        );
        RCLCPP_INFO(this->get_logger(), "Action server has been started.");
    }

// you get a unique identifier for the goal and the goal itself (to be precise, this is a const shared pointer to the goal).
private:
    // Every new received goal will be processed here first
    // We can decide to accept or reject the incoming goal
    rclcpp_action::GoalResponse goalCallback(
        const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const CountUntil::Goal> goal)
    {
        (void)uuid;
        RCLCPP_INFO(this->get_logger(), "Received a goal");
        if (goal->target_number <= 0) {
            RCLCPP_WARN(this->get_logger(), "Rejecting the goal, target number must be positive");
            return rclcpp_action::GoalResponse::REJECT;
        }
        RCLCPP_INFO(this->get_logger(), "Accepting the goal");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    // Any cancel request will be processed here, we can accept or reject it
    rclcpp_action::CancelResponse cancelCallback(
        const std::shared_ptr<CountUntilGoalHandle> goal_handle)
    {
        (void)goal_handle;
        RCLCPP_INFO(this->get_logger(), "Received a cancel request");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    // If a goal has been accepted, it will then be executed in this callback
    // After we are done with the goal execution we set a final state and provide the result
    // You’ve probably started to get used to seeing shared pointers everywhere, and here is no exception. 
    // We don’t create a result object, but a shared pointer to a result object.
    // In Python, we would set the goal’s final state first, and then return the result. In C++, we don’t return anything (note the void return type). 
    // We send the result at the same time as setting the goal state.
    void executeCallback(const std::shared_ptr<CountUntilGoalHandle> goal_handle)
    {
        int target_number = goal_handle->get_goal()->target_number;
        double delay = goal_handle->get_goal()->delay;
        auto result = std::make_shared<CountUntil::Result>();
        int counter = 0;

        // to handle the waiting time between each count iteration, we use a rclcpp::Rate object
        // In this rate object, we have to pass the rate—that is, the frequency we want for the loop. For example, if the delay is 0.5 seconds, the frequency would be 2.0 Hz.
        rclcpp::Rate loop_rate(1.0/delay);

        RCLCPP_INFO(this->get_logger(), "Executing the goal");
        for (int i = 0; i < target_number; i++) {
            counter++;
            RCLCPP_INFO(this->get_logger(), "%d", counter);
            loop_rate.sleep();
        }

        result->reached_number = counter;
        goal_handle->succeed(result);
    }

    rclcpp_action::Server<CountUntil>::SharedPtr count_until_server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}